
```{r load_packages}
library(patchwork)
library(Seurat)
library(Matrix)
options(stringsAsFactors = FALSE)
```
# load Seurat object
# for seurat version < 5.0.0 normalized data: seurat_object[["RNA"]]@data # normalized data matrix
```{r load_seurat}
seurat_obj <- readRDS(file="/w5home/bmoore/Pierre_sc_zebrafish/SCI6WK_PARSE_PG_SB_030123.rds")
print(seurat_obj)
```
# get data matrix from seurat
```{r get_input}
data.input <- seurat_obj[["RNA"]]$data
print(head(data.input))

head(seurat_obj@meta.data)
table(seurat_obj$seurat_clusters)
# export metadata
seurat.obj.zebrafish.metadata<- as.data.frame(seurat_obj@meta.data)
write.table(seurat.obj.zebrafish.metadata, file = paste0(output,"zebrafish_metadata.txt"), sep="\t", 
            quote=F, row.names = T)

# get complete metadata
seurat.obj.zebrafish.metadata <-na.omit(seurat.obj.zebrafish.metadata)
dim(seurat.obj.zebrafish.metadata)

```
# get orthologs
```{r subset_orthos}
ortholog_file <-"/w5home/bmoore/Pierre_sc_zebrafish/zebrafish_homologs_ensembl.txt"
orthologs <- read.csv2(ortholog_file, sep = "\t", header = TRUE)
print(head(orthologs))
orthologs <- orthologs[,c("Gene.stable.ID","Gene.name","Mouse.gene.name")]
```
# merge expr df with orthologs
```{r merge_expression_orthologs}
# Convert expression matrix to dataframe with gene names as a column
expr_df <- as.data.frame(as.matrix(data.input))
expr_df$gene_name <- toupper(rownames(expr_df))

# Convert ortholog gene names to uppercase
orthologs$Gene.name <- toupper(orthologs$Gene.name)
#orthologs$Mouse.gene.name <- toupper(orthologs$Mouse.gene.name)
orthologs$Gene.stable.ID <- toupper(orthologs$Gene.stable.ID)

# Create two separate ortholog dataframes and remove duplicates from each
# For Gene.name matches
orthologs_gene <- orthologs[!duplicated(orthologs$Gene.name), ]
print(paste0("Number of unique Gene.name entries: ", nrow(orthologs_gene)))

# For Gene.stable.ID matches
orthologs_stable <- orthologs[!duplicated(orthologs$Gene.stable.ID), ]
print(paste0("Number of unique Gene.stable.ID entries: ", nrow(orthologs_stable)))

# Create two separate merges - one for each potential match column
merge1 <- merge(expr_df, orthologs_gene, by.x = "gene_name", by.y = "Gene.name", all.x = TRUE)
merge2 <- merge(expr_df, orthologs_stable, by.x = "gene_name", by.y = "Gene.stable.ID", all.x = TRUE)

# Print number of matches from each merge
print(paste0("Number of matches from Gene.name: ", sum(!is.na(merge1$Gene.stable.ID))))
print(paste0("Number of matches from Gene.stable.ID: ", sum(!is.na(merge2$Gene.name))))

# Combine the results, prioritizing matches from merge1 if they exist
missing_in_merge1 <- is.na(merge1$Gene.stable.ID)
matches_in_merge2 <- !is.na(merge2$Gene.name)

# Replace missing matches with those found in merge2
final_merged <- merge1
final_merged[missing_in_merge1 & matches_in_merge2,] <- merge2[missing_in_merge1 & matches_in_merge2,]

# Check how many genes were matched in total
n_matched <- sum(!is.na(final_merged$Mouse.gene.name))
n_total <- nrow(final_merged)
print(paste0("Total matched mouse genes after combining both approaches: ", 
        n_matched, " out of ", n_total, " genes"))

# Print a few examples of the matches
print("Sample of first few matches:")
print(head(final_merged[!is.na(final_merged$Mouse.gene.name), 
        c("gene_name", "Gene.stable.ID", "Mouse.gene.name")]))
out_orthos <- final_merged[,c("gene_name", "Gene.stable.ID", "Mouse.gene.name")]
write.csv2(out_orthos, 
        file = "/w5home/bmoore/Pierre_sc_zebrafish/merged_orthologs_zebrafish-mouse.csv", 
        row.names=TRUE)
print(dim(final_merged))
```
# subset orthologs
```{r subset_orthos}
# remove mouse genes that are NA or empty
final_merged <- final_merged[!is.na(final_merged$Mouse.gene.name) & final_merged$Mouse.gene.name != "",]

# remove other gene names
data.input.matched <- final_merged[, !colnames(final_merged) %in% c("gene_name", "Gene.stable.ID")]
# make numeric
data.input.matched <- as.data.frame(data.input.matched)
data.input.matched <- mutate_all(data.input.matched[, !colnames(data.input.matched) %in% 
                        c("Mouse.gene.name")], function(x) as.numeric(as.character(x)))
data.input.matched$Mouse.gene.name <- final_merged$Mouse.gene.name
# merge duplicate mouse genes
data.input.matched <- aggregate(data.input.matched, 
                        by = list(Mouse.gene.name = data.input.matched$Mouse.gene.name), 
                        FUN = mean)
print(dim(data.input.matched))
# write
write.csv2(data.input.matched, 
        file="/w5home/bmoore/Pierre_sc_zebrafish/mouse_ortho_sc_expr_matrix.csv", 
        row.names=TRUE)
print(colnames(data.input.matched))
```

```{r format}
# read expression data back in with mouse
data.input.matched <- read.csv2(file="/w5home/bmoore/Pierre_sc_zebrafish/mouse_ortho_sc_expr_matrix.csv", 
        row.names=1)
print(colnames(data.input.matched))
# remove last column that has Mouse.gene.name (as this is duplicated)
data.input.matched2 <- data.input.matched[,-61487]
dim(data.input.matched2)
# Convert all numeric columns to numeric type
numeric_cols <- colnames(data.input.matched2)[-1]  # all columns except mouse name
head(numeric_cols)
# Set row names from Group.1 and remove that column
rownames(data.input.matched2) <- data.input.matched2$Mouse.gene.name
data.input.matched2 <- data.input.matched2[,numeric_cols]
data.input.matched2[1:5,1:5]
# make numeric
#data.input.matched <- as.data.frame(data.input.matched)
#data.input.matched <- mutate_all(data.input.matched, function(x) as.numeric(as.character(x)))

# Convert to matrix and then sparse matrix
data.input.matched2 <- as.matrix(data.input.matched2)
data.input.matched2 <- Matrix(data.input.matched2, sparse = TRUE)
print(dim(data.input.matched2))

```
# convert back to orthologous mouse seurat obj
```{r convert2seurat}
# read complete metadata
seurat.obj.zebrafish.metadata <- read.table(file = paste0(output,"zebrafish_metadata.txt"), sep="\t", 
            row.names = T)
# convert to seurat obj
seurat.obj.mouse <- CreateSeuratObject(
  counts=data.input.matched2,
  assay = "RNA",
  meta.data = seurat.obj.zebrafish.metadata, #meta
  project = "zebrafish-mouse-spine-regen")
print(seurat.obj.mouse)

# data was already normalized, so add back in counts as data layer
seurat.obj.mouse[["RNA"]]$data <- seurat.obj.mouse[["RNA"]]$counts
print(seurat.obj.mouse)
# check out metadata
m_metadata <- seurat.obj.mouse[[]]
colnames(m_metadata)
unique(m_metadata$customclassif)
```
```{r run_dim_red}
all.genes <- rownames(seurat.obj.mouse)
seurat.obj.mouse <- ScaleData(seurat.obj.mouse, features = all.genes)
seurat.obj.mouse <- FindVariableFeatures(seurat.obj.mouse)
seurat.obj.mouse <- RunPCA(seurat.obj.mouse, features = VariableFeatures(object = seurat.obj.mouse))
seurat.obj.mouse <- RunUMAP(seurat.obj.mouse, dims = 1:20)
```
```{r reannotate}
# set idents to cluster value
seurat.obj.mouse <- SetIdent(seurat.obj.mouse, value = "seurat_clusters")
# read in new annotation
annot_df <- read.csv(file="/w5home/bmoore/Pierre_sc_zebrafish/celltype_annot_df.csv")
# order annot_df
annot_df.ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
# get new cluster ids
new.cluster.ids <- annot_df.ordered$Celltype 
# Rename the clusters based on the new IDs
names(new.cluster.ids) <- levels(seurat.obj.mouse)
seurat.obj.mouse <- RenameIdents(seurat.obj.mouse, new.cluster.ids)
# put in CellType metadata
seurat.obj.mouse$CellType <- Idents(seurat.obj.mouse)

# Generate and plot the UMAP plot
pdf(paste0(output, "new-labeled-clusters.pdf"), bg = "white")
print(DimPlot(seurat.obj.mouse, reduction = "umap", label = TRUE, 
        pt.size = 0.5, group.by = "CellType"))
dev.off()
```
# subset mouse by sampel type
```{r subset}
table(seurat.obj.mouse$Type)
# 1 week
seurat.obj.mouse_1wk <- subset(x = seurat.obj.mouse, subset = Type == "1Wk")
print(seurat.obj.mouse_1wk)
print(table(seurat.obj.mouse_1wk$CellType))
# 2 week non recover
seurat.obj.mouse_2WkNR <- subset(x = seurat.obj.mouse, subset = Type == "2Wk-NR")
print(seurat.obj.mouse_2WkNR)
print(table(seurat.obj.mouse_2WkNR$CellType))
# 2 week recover
seurat.obj.mouse_2WkR <- subset(x = seurat.obj.mouse, subset = Type == "2Wk-R")
print(seurat.obj.mouse_2WkR)
print(table(seurat.obj.mouse_2WkR$CellType))
# 4 week non recover
seurat.obj.mouse_4WkNR <- subset(x = seurat.obj.mouse, subset = Type == "4Wk-NR")
print(seurat.obj.mouse_4WkNR)
print(table(seurat.obj.mouse_4WkNR$CellType))
# 4 week recover
seurat.obj.mouse_4WkR <- subset(x = seurat.obj.mouse, subset = Type == "4Wk-R")
print(seurat.obj.mouse_4WkR)
print(table(seurat.obj.mouse_4WkR$CellType))
# 6 week non recover
seurat.obj.mouse_6WkNR <- subset(x = seurat.obj.mouse, subset = Type == "6Wk-NR")
print(seurat.obj.mouse_6WkNR)
print(table(seurat.obj.mouse_6WkNR$CellType))
# 6 week recover
seurat.obj.mouse_6WkR <- subset(x = seurat.obj.mouse, subset = Type == "6Wk-R")
print(seurat.obj.mouse_6WkR)
print(table(seurat.obj.mouse_6WkR$CellType))

```
# save objects
```{r save}
output <- "/w5home/bmoore/Pierre_sc_zebrafish/"
saveRDS(seurat.obj.mouse_1wk, file=paste0(output,"seurat_mouse_annot_1wk.rds"))
saveRDS(seurat.obj.mouse_2WkNR, file=paste0(output,"seurat_mouse_annot_2wkNR.rds"))
saveRDS(seurat.obj.mouse_2WkR, file=paste0(output,"seurat_mouse_annot_2wkR.rds"))
saveRDS(seurat.obj.mouse_4WkNR, file=paste0(output,"seurat_mouse_annot_4wkNR.rds"))
saveRDS(seurat.obj.mouse_4WkR, file=paste0(output,"seurat_mouse_annot_4wkR.rds"))
saveRDS(seurat.obj.mouse_6WkNR, file=paste0(output,"seurat_mouse_annot_6wkNR.rds"))
saveRDS(seurat.obj.mouse_6WkR, file=paste0(output,"seurat_mouse_annot_6wkR.rds"))
saveRDS(seurat.obj.mouse, file=paste0(output,"seurat_mouse_annot.rds"))
# save with qs
#install.packages("qs2")
#remotes::install_cran("qs2", type = "source", configure.args = "--with-TBB --with-simd=AVX2")
library(qs2)
qs_save(seurat.obj.mouse, "seurat_mouse_annot.qs2")
# convert to rds
qs_to_rds(input_file = "seurat_mouse_annot.qs2", output_file = "seurat_mouse_annot.rds")
# read back in
#data <- qs_read("myfile.qs2", validate_checksum = TRUE)
# or
# xrds <- readRDS(file_rds)
```