---
title: "preprocess_for_cross-species"
author: "Beth Moore"
date: "2023-07-07"
output: html_document
---
# Preprocessing for cross species analysis
# to have cross species, we need ortholog genes
# we then need to subset both reference and query by these genes
# reference data will also be subseted by its metadata

# Load libraries, set wd, and source functions
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
# #install.packages("Matrix", repos="http://R-Forge.R-project.org")
library(Matrix)
# set up dirs
WD <- "/w5home/bmoore/scRNAseq/GAMM/human_data/cowan_cell_2020/"
GIT_DIR <- "/w5home/bmoore/Gamm_scRNAseq/"
REF.MATRIX <- "STAR_output/output_F49i-N-B7/Solo.out/GeneFull/"
QUERY.MATRIX <- "/w5home/bmoore/scRNAseq/GAMM/STAR_output/output_S2_mm_mt/Solo.out/GeneFull/"
REF.PROJ.NAME <- "F49i-N-B7"
QUERY.PROJ.NAME <- "gamm.data.S2"
# use environment
use_condaenv(condaenv="/w5home/bmoore/miniconda3/envs/scRNAseq_best/")
# set working directory
setwd(WD)
# create output folder
#timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
#output <- paste0("output_preprocess_", timestamp)
#dir.create(output, showWarnings = FALSE)
#output <- paste0(output, "/")
output <- "/w5home/bmoore/scRNAseq/GAMM/human_data/cowan_cell_2020/output_preprocess_20240229_101646/"
# copy config and source functions
#file.copy(paste0(GIT_DIR,"config.json"), file.path(paste0(output, "config.json")))
config <- fromJSON(file.path(output, "config.json"))
source(paste0(GIT_DIR,"src/sc_pipeline_functions.R"))

```
# Read in data
```{r data}
# refernce data
ref.matrix <- read_aligned_data(paste0(WD,REF.MATRIX),REF.PROJ.NAME, output)

# load query data
query.matrix <- read_aligned_data(QUERY.MATRIX,QUERY.PROJ.NAME, output)
```

# run soupx
```{r soupx}
all_aligned_data <- list(ref.matrix,query.matrix)
soupX_objs <- purrr::map(all_aligned_data, ~prep_seurat_and_soupX(data.raw = .x$raw, data = .x$filtered, project = .x$project), .names = purrr::map_chr(all_data, 'project'))
# if umap-learn error try: reticulate::py_install(packages ='umap-learn')
# get soup matrices
ref.matrix.soupx<- soupX_objs[[1]]$out
query.matrix.soupx<- soupX_objs[[2]]$out
```
# Create Seurat objects and convert to sce objects for doublet removal
```{r}
project_names <- purrr::map_chr(all_aligned_data, 'project')
feature_set1 <- list(feature1 = 'nCount_RNA', feature2 = 'nFeature_RNA')
filtered_list_of_objs <- purrr::map2(soupX_objs, project_names, ~create_seurat_and_sce(out = .x$out, project = .y, feature_set = feature_set1))
print(project_names)
```
# Doublet removal and lane merging
```{r}
combine_feature_plots(filtered_list_of_objs, feature_set1, file_name_base = "post_soupx_qc")
filtered_sce_list <- list(filtered_list_of_objs[[1]]$sce, filtered_list_of_objs[[2]]$sce)
lane_and_merged_seurat_obj <- run_scDblFinder_and_merge(filtered_sce_list)
print(unique(lane_and_merged_seurat_obj$orig.ident))
```
# get sparse count matrices
```{r}
obj.list <- SplitObject(lane_and_merged_seurat_obj, split.by = "orig.ident")
human_ref<- obj.list[[1]]
pig_query<- obj.list[[2]]
ref.matrix.flt <- human_ref@assays$RNA@data
gam.matrix.flt <- pig_query@assays$RNA@data

rm(ref.matrix.soupx,query.matrix.soupx)
```
# create seurat object
```{r} 
human_ref.seurat<- CreateSeuratObject(ref.matrix.flt, project = project_names[[1]], assay = "RNA",
                                       min.cells = 3, min.features = 200)
pig_query.seurat<- CreateSeuratObject(gam.matrix.flt, project = project_names[[2]], assay = "RNA",
                                       min.cells = 3, min.features = 200)
```
# Mitochondrial gene filtering
```{r}
human_ref.seurat_mt_filtered <- filter_cells(human_ref.seurat)
pig_query.seurat_mt_filtered <- filter_cells(pig_query.seurat)
```

# Normalization
```{r normalization}
ref.seurat_norm <- normalize_data(human_ref.seurat_mt_filtered)
query.seurat_norm <- normalize_data(pig_query.seurat_mt_filtered)
saveRDS(ref.seurat_norm, file = paste0(output,"human_",project_names[[1]],"_norm.rds"))
saveRDS(query.seurat_norm, file = paste0(output,"query_",project_names[[2]],"_norm.rds"))
rm(human_ref.seurat,pig_query.seurat)
```

# Feature selection
```{r feat_select}
#gamm.S2.seurat_norm<-readRDS(file = "output_20230711_155556/GAMM_S2_norm.rds")
#human_D205.seurat_norm<-readRDS(file = "output_20230711_155556/human_D205_norm.rds")
query.seurat_feat_select <- feature_selection(query.seurat_norm)
ref.seurat_feat_select <- feature_selection(ref.seurat_norm)
rm(human_ref.seurat_mt_filtered,pig_query.seurat_mt_filtered)
```
# Scaling data
```{r scale}
#query.seurat_scaled <- scale_data(query.seurat_feat_select)
ref.seurat_feat_select<- readRDS(file= paste0(output,"human_F49i-N-B7_featsel.rds"))
query.seurat_scaled<- readRDS(file=paste0(output,"query_gamm.data.S2_scaled.rds"))
ref.seurat_scaled <- scale_data(ref.seurat_feat_select)
#rm(ref.seurat_norm,query.seurat_norm)
```
# Dimensional reduction
```{r dim_reduct}
query.seurat_dim_reduced <- run_and_visualize_pca(query.seurat_scaled)
ref.seurat_dim_reduced <- run_and_visualize_pca(ref.seurat_scaled)
```
# add in batch correction
```{r batch_corr}
query.seurat_batch_corr <- perform_batch_correction(query.seurat_dim_reduced)
ref.seurat_batch_corr <- perform_batch_correction(ref.seurat_dim_reduced)
```
# grab variable features
```{r}
# get variable genes from seurat:
#FEARURE_LIST <- query.seurat@var.genes
# or variable genes from seurat 3
feature_list_Q <- query.seurat_dim_reduced@assays$RNA@var.features
feature_list_Q <- as.vector(feature_list_Q)
feature_list_R <- ref.seurat_dim_reduced@assays$RNA@var.features
feature_list_R <- as.vector(feature_list_R)
print(feature_list_Q)
print(feature_list_R)
```
```{r}
# get scale data
scaled_matrix_Q <- query.seurat_dim_reduced@assays$RNA@scale.data
scaled_matrix_R <- ref.seurat_dim_reduced@assays$RNA@scale.data
```
# save seurat object
```{r}
saveRDS(ref.seurat_dim_reduced, file = paste0(output,"human_",project_names[[1]],"_dim_reduced.rds"))
saveRDS(query.seurat_dim_reduced, file = paste0(output,"query_",project_names[[2]],"_dim_reduced.rds"))
rm(human_ref.seurat_mt_filtered,human_ref.seurat,pig_query.seurat_mt_filtered,pig_query.seurat)
```
# do ortho subsetting
# read in orthologs and subset
```{r ortho_subset}
project_names <- list("F49i-N-B7","gamm.data.S2")
objs.list <- ortholog_subset(ref.seurat_dim_reduced,query.seurat_dim_reduced,project_names)
ref.seurat <- objs.list[[1]]
query.seurat <- objs.list[[2]]
orthologs <- objs.list[[3]]
print(ref.seurat)
print(query.seurat)
```
# get metadata
```{r metadata}
# "f49b7"
ref.seurat <- get_metadata(ref.seurat, "ref")

print(colnames(ref.seurat@meta.data))
print(table(ref.seurat$orig.ident))
print(table(ref.seurat$EGA))
print(table(ref.seurat$cell_type2))
```
# add back in variable features from previous object
```{r}
# add variable feature back in
VariableFeatures(query.seurat) <- feature_list_Q
VariableFeatures(ref.seurat) <- feature_list_R
```
```{r}
# add scaled data back in
query.seurat@assays$RNA@scale.data <- scaled_matrix_Q 
ref.seurat@assays$RNA@scale.data <- scaled_matrix_R
```
```{r}
# redo PCA
query.seurat_dim_reduced <- run_and_visualize_pca(query.seurat)
ref.seurat_dim_reduced <- run_and_visualize_pca(ref.seurat)
```
# Run umap
```{r umap}
query.seurat_umap <- run_umap(query.seurat_dim_reduced)
ref.seurat_umap <- run_umap(ref.seurat_dim_reduced)
```
# clustering
```{r cluster}
query.seurat_clustered <- perform_clustering(query.seurat_umap)
ref.seurat_clustered <- perform_clustering(ref.seurat_umap)

```
# save
```{r save}
# save objects
saveRDS(query.seurat_clustered, file = paste0(output, "query",project_names[[2]],"_clustered.rds"))
saveRDS(ref.seurat_clustered, file = paste0(output, "human_",project_names[[1]],"_clustered.rds"))
# save session info
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```