---
title: "recluster and annotate gamm clusters"
output: html_document
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
library(ggplot2)
library(scater)
library(SingleCellExperiment)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/" #"/Users/bmoore/Desktop/GitHub/scRNAseq/GAMM/"
use_condaenv("scRNAseq_best")
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/reh_cellrep_2020/")) #"GAMM_S2/"
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json")) # "/Users/bmoore/Desktop/GitHub/scRNAseq_library/config.json"
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_library/src/sc_pipeline_functions.R") # "/Users/bmoore/Desktop/GitHub/scRNAseq_library/src/sc_pipeline_functions.R"
```
# Load Data
```{r load_data}
gamms2<- readRDS(file = "GAMM_S2_dimRed-harmony-clusters.rds")
```
# recluster
```{r recluster}

num_replicate = config$perform_clustering$num.replicate
  dims = 1:config$perform_clustering$dims
  dims_umap = 1:config$perform_clustering$dims_umap
  resolution = config$perform_clustering$resolution
  algorithm = config$perform_clustering$algorithm
  umap.method = config$perform_clustering$umap.method
  
  pdf(paste0(output, "elbow_pca.pdf"), width = 8, height = 6)
  elbow_pca <- ElbowPlot(gamms2, reduction='pca')
  print(elbow_pca)
  dev.off()
  
  gamms2_recluster <- FindNeighbors(gamms2, dims = dims_umap, reduction='harmony')
  gamms2_recluster <- FindClusters(gamms2_recluster, resolution = resolution, algorithm = algorithm)
```
# visualize
```{r visualize_clusters}
pdf(paste0(output, "umap_clusters.pdf"), width = 8, height = 6)
  umap_clusters <- DimPlot(gamms2_recluster, reduction = "umap", label = TRUE, pt.size = .1)
  print(umap_clusters)
  dev.off()
```
# save object
```{r save}
saveRDS(gamms2_recluster, file = "GAMM_S2_recluster_res0.14.rds")
```
# Find markers with Differential expressed features (genes)
```{r de_genes}
de_results1 <- find_differentially_expressed_features(gamms2)
de_results2 <- find_differentially_expressed_features(gamms2_recluster)
```
# analyse known markers
# note: results overwrite
```{r analyze_known_markers}
known_markers="/Users/bmoore/Desktop/GitHub/scRNAseq/GAMM/Gamm_lab_Consolidated_markerList.txt"
de.markers=de_results1[[1]]
analyze_known_markers(gamms2,de.markers,known_markers,output)
de.markers=de_results2[[1]]
analyze_known_markers(gamms2_recluster,de.markers,known_markers,output)
```
# analyze known markers 2
# note: results overwrite
```{r analyze_known_markers2}
score_and_plot_markers(gamms2,known_markers,output)
score_and_plot_markers(gamms2_recluster,known_markers,output)
```
# manual annotation
```{r manual_annot}
# cluster ids for gamm S2 at 0.5 with consolidated markers (for clusters 1-18)
new.cluster.ids1 <- c("Retinal Progenitor-Muller Glia","Pan PRs","Pan PRs-Rods","Pan PRs-Rods","Pan PRs-Rods","Pan PRs-Rods","Ganglion cell","Bipolar Cells-Ganglion cell","Pan PRs-Rods","unknown","Pan PRs","Cones-Pan PRs","unknown","unknown","Muller Glia-PanPRs-Rods","unknown","Muller Glia","Bipolar Cells")
# annotate
gamms2<-annotate_clusters_and_save(gamms2, new.cluster.ids1, output)
# cluster ids for gamm S2 at 0.14 with consolidated markers (for clusters 1-8)
new.cluster.ids2 <- c("Pan PRs-Rods","Muller Glia","Ganglion cell","Bipolar Cells","Pan PRs-Rods",
                      "Rods-Muller Glia-Pan PRs","unknown","Cones-Pan PRs")
# annotate
gamms2_recluster<-annotate_clusters_and_save(gamms2_recluster, new.cluster.ids2, output)
```
# save session info
```{r sessioninfo}
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```