---
title: "recluster and annotate clusters"
output: html_document
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
library(ggplot2)
library(scater)
library(SingleCellExperiment)
# set variables
GIT_DIR <- "/w5home/bmoore/scRNAseq_downstream/"
config <- fromJSON(file.path(GIT_DIR, "config.json"))
WD <- config$recluster$DATA_DIR
SEURAT.FILE <- config$recluster$SEURAT.FILE
# set up environment and output
use_condaenv("/w5home/bmoore/miniconda3/envs/scRNAseq_best", required=TRUE)
setwd(WD)
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_recluster_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy(file.path(GIT_DIR, "config.json"), file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_downstream/src/sc_pipeline_functions.R") 
```
# Load Data
```{r load_data}
seurat.obj <- readRDS(file = SEURAT.FILE)
```
# recluster
```{r recluster}
seurat.obj_recluster <- perform_clustering(seurat.obj, output, "recluster")
```
# save object
```{r save}
resolution <- as.character(config$recluster$perform_clustering$resolution)
saveRDS(seurat.obj_recluster, file = paste0(output, "seuratobj_recluster_res", resolution, ".rds"))
```
# Find markers with Differential expressed features (genes)
# analyse known markers
# note: results overwrite
```{r analyze_known_markers}
annot_df <- score_and_plot_markers(seurat.obj_recluster,output,"recluster")
```
# manual annotation
```{r manual_annot}
# order annot_df
annot_df.ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
new.cluster.ids <- annot_df.ordered$Cell.type
# annotate
seurat.obj_recluster<-annotate_clusters_and_save(seurat.obj_recluster, new.cluster.ids, output)
# print table of celltypes
print(table(seurat.obj_recluster@meta.data$CellType1))
```
# save annotation as dataframe
```{r save_annot}
seurat.obj_recluster.metadata<- as.data.frame(seurat.obj_recluster@meta.data)
write.table(seurat.obj_recluster.metadata, file = paste0(output,"manual_annot_metadata_",resolution,".txt"), sep="\t", 
            quote=F, row.names = T)
```
# save session info
```{r sessioninfo}
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```
# optional- 
# read back in metadata
```{r get_meta}
# # add in metadata from previous analysis
# metadata <- read.csv(paste0(output,"manual_annot_metadata_",resolution,".txt"), 
#             row.names = 1, header = TRUE, sep = "\t")

# # add metadata to query data
# seurat.obj_recluster <- AddMetaData(seurat.obj_recluster, metadata)

# # check again
# colnames(seurat.obj_recluster@meta.data)
# table(seurat.obj_recluster@meta.data$CellType1)
```