---
title: "recluster and annotate clusters"
output: html_document
---

# ENVIRONMENT SETUP

```{r env, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(scran)
library(BiocParallel)
library(DropletUtils)
library(cowplot)
library(harmony)
library(SoupX)
library(scDblFinder)
library(reticulate)
library(purrr)
library(jsonlite)
library(rmarkdown)
library(ggplot2)
library(scater)
library(SingleCellExperiment)
# set variables
WD <- "/w5home/bmoore/scRNAseq/GAMM/GAMM_S2/"
SEURAT.FILE <- "output_20230830_155530/seurat_obj_labeled.rds"
GIT_DIR <- "/w5home/bmoore/scRNAseq_downstream/"
# set up environment and output
use_condaenv("/w5home/bmoore/miniconda3/envs/scRNAseq_best", required=TRUE)
setwd(WD)
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_recluster_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_downstream/config.json", file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_downstream/src/sc_pipeline_functions.R") 
```
# Load Data
```{r load_data}
seurat.obj <- readRDS(file = SEURAT.FILE)
```
# recluster
```{r recluster}
seurat.obj_recluster <- perform_clustering (seurat.obj, output)
```
# save object
```{r save}
saveRDS(seurat.obj_recluster, file = paste0(SEURAT.FILE, "_recluster_res", resolution, ".rds"))
```
# Find markers with Differential expressed features (genes)
# analyse known markers
# note: results overwrite
```{r analyze_known_markers}
annot_df <- score_and_plot_markers(seurat.obj_recluster,output)
```
# manual annotation
```{r manual_annot}
# order annot_df
annot_df.ordered <- annot_df[order(as.numeric(annot_df$Cluster)), ]
new.cluster.ids <- annot_df.ordered$Cell.type
# annotate
gseurat.obj_recluster<-annotate_clusters_and_save(seurat.obj_recluster, new.cluster.ids, output)

```
# save session info
```{r sessioninfo}
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```