---
title: "sc Pred"
author: "Beth Moore"
date: "2023-09-18"
output: html_document
---
# load packages
```{r load_packages}
knitr::opts_chunk$set(echo = TRUE)
# conda activate scRNAseq_best
#devtools::install_github("powellgenomicslab/scPred")
library(scPred)
library(Seurat)
library(magrittr)
library(harmony)
library(rmarkdown)
library(jsonlite)
library(purrr)
library(scran)
library(patchwork)
library(dplyr)
library(reticulate)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/"
METADATA <- "gamm_manual_annot_metadata_c0.5.txt" # add to working directory
use_condaenv(condaenv = 'scRNAseq_best', required = TRUE)
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/cowan_cell_2020/"))
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_scPred_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_library/sc_pipeline/src/sc_pipeline_functions.R")
```
# run preprocess_crossspecies.Rmd first to get the data preprocessed the same way
# get reference and query data
```{r get_data}
#human_F49B7.seurat <- readRDS(file = "output_seurat_mapping_20231207_103331/human_F49B7_cca_pred.rds")
#human_IMR90.seurat <- readRDS(file = "output_seurat_mapping_20231207_103331/human_IMR90_cca_pred.rds")
#gamms2.1 <- readRDS(file = "output_seurat_mapping_20231207_103331/gamms2.1_cca_pred.rds")
#gamms2.2 <- readRDS(file = "output_seurat_mapping_20231207_103331/gamms2.2_cca_pred.rds")
human_F49B7.seurat <- readRDS(file = "output_preprocess20231130_141317/human_F49B7_clustered.rds")
human_IMR90.seurat <- readRDS(file = "output_preprocess20231130_141317/human_IMR90_clustered.rds")
gamms2 <- readRDS(file = "output_preprocess20231130_141317/gamm.S2_clustered.rds")
```
# subset metadata if needed
```{r subset_metadata}
# read in meta.data
# remember the set working directory in chunk 1, to imform this path
metadata <- read.csv2("meta_30weeks_mod.txt", sep="\t", header=TRUE, row.names=1)
# subset by D205 data
metadata1 <- subset(metadata, metadata$source=="f49b7")
metadata1<- as.data.frame(metadata1)
complete.cases(metadata1)
metadata1<-na.omit(metadata1)
metadata2 <- subset(metadata, metadata$source=="imr90")
metadata2<- as.data.frame(metadata2)
complete.cases(metadata2)
metadata2<-na.omit(metadata2)
# get numbers for each celltype
table(metadata1$cell_type3)
table(metadata2$cell_type3)
human_F49B7.seurat<-AddMetaData(object=human_F49B7.seurat, metadata=metadata1)
human_IMR90.seurat<-AddMetaData(object=human_IMR90.seurat, metadata=metadata2)
# check that it got incorporated into the seurat object
table(human_F49B7.seurat@meta.data$cell_type3)
table(human_IMR90.seurat@meta.data$cell_type3)
# check metadata
table(human_F49B7.seurat$cell_type2)
table(human_IMR90.seurat$cell_type2)
# subset if needed
human_IMR90.seurat<- subset(x = human_IMR90.seurat, subset = cell_type3 != "Ast")

table1 <- table(human_IMR90.seurat@meta.data$cell_type2)
write.table(table1, file = paste0(output,"IMR90_celltype_table.txt"), sep="\t", row.names = FALSE)
# check metadata
table(human_F49B7.seurat@meta.data$cell_type2)
# subset if needed
human_F49B7.seurat<- subset(x = human_F49B7.seurat, subset = cell_type2 != "AC-MC")
human_F49B7.seurat<- subset(x = human_F49B7.seurat, subset = cell_type2 != "CdBC-MC")
human_F49B7.seurat<- subset(x = human_F49B7.seurat, subset = cell_type2 != "L/M cone-rod")
human_F49B7.seurat<- subset(x = human_F49B7.seurat, subset = cell_type2 != "RPE")

table2<- table(human_F49B7.seurat@meta.data$cell_type2)
write.table(table2, file = paste0(output,"F49B7_celltype_table.txt"), sep="\t", row.names = FALSE)
```
```{r subset_test_set_function}

get_train_test_set <- function(seurat_obj, annot){
  # Set the percentage of cells to sample (10% in this case)
  sampling_percentage <- 0.1
  # Calculate the number of cells to sample for each cell type
  cell_types <- unique(annot)
  cells_to_sample <- ceiling(table(annot) * sampling_percentage)

  # Create a copy of the original Seurat object to keep the unsubsetted cells
  seurat_obj.train <- seurat_obj
  # Initialize an empty list to store subsetted Seurat objects
  subset_list <- list()

  # Iterate through each cell type, subset the cells, and store the subset in the list
  for (cell_type in cell_types) {
    cell_indices <- which(annot == cell_type)
    sampled_indices <- sample(cell_indices, cells_to_sample[cell_type], replace = FALSE)
  
    subset_list[[cell_type]] <- seurat_obj[, sampled_indices]
    # Remove the subsetted cells from the original Seurat object
    seurat_obj.train <- seurat_obj.train[, -sampled_indices]
  }

# Create a new Seurat object with the subsetted cells
  seurat_obj.test <- CreateSeuratObject(counts = subset_list[[1]]@assays$RNA@counts,
                                           meta.data = subset_list[[1]]@meta.data)
  for (i in 2:length(subset_list)) {
    seurat_obj.test <- merge(seurat_obj.test, subset_list[[i]])
  }
  seurat_obj.test$Idents <- annot

  obj_vector <- c(seurat_obj.train,seurat_obj.test)
  return(obj_vector)
}


```
# get test set
```{r test}
annot1 <- human_F49B7.seurat$cell_type2
objv1 <- get_train_test_set(human_F49B7.seurat,annot1)
print(objv1)
human_F49B7.seurat.train <- objv1[[1]]
human_F49B7.seurat.test <- objv1[[2]]
print(human_F49B7.seurat.train)
print(human_F49B7.seurat.test)
```
```{r}
annot2 <- human_IMR90.seurat$cell_type3
objv2 <- get_train_test_set(human_IMR90.seurat,annot2)
print(objv2)
human_IMR90.seurat.train <- objv2[[1]]
human_IMR90.seurat.test <- objv2[[2]]
print(human_IMR90.seurat.train)
print(human_IMR90.seurat.test)
```
# Perform further analysis as needed on the seurat_test_object
```{r normalize_test_set}
human_F49B7.seurat.test_norm <- normalize_data(human_F49B7.seurat.test)
human_IMR90.seurat.test_norm <- normalize_data(human_IMR90.seurat.test)
#saveRDS(human_D205.seurat.test_norm, file= paste0(output,"human_D205_test_norm.rds"))
```

# do analysis on training set- don't do if already done!
```{r redo_analysis}
# no need to normalize again
#feature selection
human_D205.seurat.train <- feature_selection(human_D205.seurat.train)
# scale
human_D205.seurat.train <- scale_data(human_D205.seurat.train)
# run pca
human_D205.seurat.train <- run_and_visualize_pca(human_D205.seurat.train)
# run umap
human_D205.seurat.train <- run_umap(human_D205.seurat.train)
#save
#saveRDS(human_D205.seurat.train, file= paste0(output,"human_D205_train.rds"))
```
# show umap of ref data
```{r show_umap_ref}
# we don't do clustering because clusters already annotated
 # Save UMAP clusters plot
  pdf(paste0(output, "F49B7_umap_ref.pdf"), width = 8, height = 6)
  umap_clusters1 <- DimPlot(human_F49B7.seurat.train, group.by = "cell_type2", 
  reduction = "umap", label = TRUE, pt.size = .1)
  print(umap_clusters1)
  dev.off()
  pdf(paste0(output, "IMR90_umap_ref.pdf"), width = 8, height = 6)
  umap_clusters2 <- DimPlot(human_IMR90.seurat.train, group.by = "cell_type3", 
  reduction = "umap", label = TRUE, pt.size = .1)
  print(umap_clusters2)
  dev.off()
  table(human_F49B7.seurat.train$cell_type2)
  table(human_IMR90.seurat.train$cell_type3)
  # removing S cone from F49B7 because < 10
  human_F49B7.seurat.train<- subset(human_F49B7.seurat.train, subset = cell_type2 != 'S cone')
  table(human_F49B7.seurat.train$cell_type2)
  table(human_F49B7.seurat.test$cell_type2)
  human_F49B7.seurat.test<- subset(human_F49B7.seurat.test, subset = cell_type2 != 'S cone')
  table(human_F49B7.seurat.test$cell_type2)
  ```
# training the classifyer
```{r train_classifyer}
# getFeatureSpace will create a scPred object stored in the @misc slot. This 
# object will contained all required information to classify cells.
human_F49B7.seurat.train <- getFeatureSpace(human_F49B7.seurat.train, "cell_type2")
human_IMR90.seurat.train <- getFeatureSpace(human_IMR90.seurat.train, "cell_type3")

# train models for each cell type
# cv has to be set BELOW min class
human_F49B7.seurat.train <- trainModel(human_F49B7.seurat.train, resampleMethod = "cv",
                                number = 10, seed = 704)
human_IMR90.seurat.train <- trainModel(human_IMR90.seurat.train, resampleMethod = "cv",
                                number = 10, seed = 704)
# if you get this error:
# There were missing values in resampled performance measures.
# try: 1. make sure no NAs in data 2. decreasing the cvs, 3. down or up sample to make classes even
```
```{r check_classifyer}
# get training prob for each cell-cell type
get_probabilities(human_F49B7.seurat.train) %>% head()
get_probabilities(human_IMR90.seurat.train) %>% head()
# use get_scpred method to retrieve the scPred object from the Seurat object.
# this gives stats on each cell type prediction
get_scpred(human_F49B7.seurat.train)
model_stats1 <- get_scpred(human_F49B7.seurat.train)
model_stats1
get_scpred(human_IMR90.seurat.train)
model_stats2 <- get_scpred(human_IMR90.seurat.train)
model_stats2
# visualize cell type probabilties
pdf(paste0(output, "celltype_modelF49B7_prob_svmradial.pdf"), width = 8, height = 6)
plot_probabilities(human_F49B7.seurat.train)
dev.off()
pdf(paste0(output, "celltype_modelIMR90_prob_svmradial.pdf"), width = 8, height = 6)
plot_probabilities(human_IMR90.seurat.train)
dev.off()
```
# OPTIONAL: can try other models from caret:
# https://topepo.github.io/caret/available-models.html
# pass in the model parameter
# can reclassify a subset of cells (like those that didn't work well)

# check test data
```{r check_test}

check_test <- function(seurat_obj.test,seurat_obj.train,name,output){
  seurat_obj.test <- scPredict(seurat_obj.test, seurat_obj.train)
  # test set in predictions space
  pdf(paste0(output, "test_celltype_prediction_",name,".pdf"), width = 8, height = 6)
  print(DimPlot(seurat_obj.test, group.by = "scpred_prediction", reduction = "scpred"))
  dev.off()
  # run UMAP using the aligned data as an input
  seurat_obj.test <- RunUMAP(seurat_obj.test, reduction = "scpred", dims = 1:30)
  # plot the predicted labels for each cell type over the UMAP:
  pdf(paste0(output, "test_celltype_prediction_",name,"-umap.pdf"), width = 8, height = 6)
  print(DimPlot(seurat_obj.test, group.by = "scpred_prediction", label = TRUE, repel = TRUE))
  dev.off()
  # compare to original labels
  pdf(paste0(output, "test_celltype_original-label_",name,"-umap.pdf"), width = 8, height = 6)
  print(DimPlot(seurat_obj.test, group.by = "cell_type2", label = TRUE, repel = TRUE))
  dev.off()

  return(seurat_obj.test)

}

human_F49B7.seurat.test <- check_test(human_F49B7.seurat.test,human_F49B7.seurat.train,"F49B7",output)
human_IMR90.seurat.test <- check_test(human_IMR90.seurat.test,human_IMR90.seurat.train,"IMR90",output)
```
# can reclassify a subset of cells (like those that didn't work well)
```{r train_classifyer2}
# other models tried: rf, logreg, xgbTree, mda, AdaBoost.M1
human_IMR90.seurat.train <- trainModel(human_IMR90.seurat.train, model = 'rf', 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("CdBC","RBC","rod","S cone"))
get_scpred(human_IMR90.seurat.train)
model_stats_ada <- get_scpred(human_IMR90.seurat.train)
pdf(paste0(output, "celltype_model-IMR90_prob_rf.pdf"), width = 8, height = 6)
plot_probabilities(human_IMR90.seurat.train)
dev.off()
# note run check test again to see if model improved on test data
human_IMR90.seurat.test <- check_test(human_IMR90.seurat.test,human_IMR90.seurat.train,"IMR90-2",output)
```
```{r train_classifyer3}
# other models 
human_F49B7.seurat.train <- trainModel(human_F49B7.seurat.train, model = "rf", 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("CdBC","ChBC","HC","rod"))
get_scpred(human_F49B7.seurat.train)
pdf(paste0(output, "celltype_model-F49B7_prob_rf.pdf"), width = 8, height = 6)
plot_probabilities(human_F49B7.seurat.train)
dev.off()
# run test data again
human_F49B7.seurat.test <- check_test(human_F49B7.seurat.test,human_F49B7.seurat.train,"F49B7-2",output)
```
# get final model
```{r train_classifyer4}
# IMR90
human_IMR90.seurat.train <- trainModel(human_IMR90.seurat.train, model = 'svmRadial', 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("CdBC","RBC","rod","S cone"))
get_scpred(human_IMR90.seurat.train)
pdf(paste0(output, "celltype_model-IMR90_prob_fin.pdf"), width = 8, height = 6)
plot_probabilities(human_IMR90.seurat.train)
dev.off()
# F49B7
human_F49B7.seurat.train <- trainModel(human_F49B7.seurat.train, model = 'svmRadial', 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("CdBC","ChBC","HC","rod"))
get_scpred(human_F49B7.seurat.train)
pdf(paste0(output, "celltype_model-F49B7_prob_fin.pdf"), width = 8, height = 6)
plot_probabilities(human_F49B7.seurat.train)
dev.off()

```
# save seurat object
```{r save_ref}
# save ref data
saveRDS(human_F49B7.seurat.train, file = paste0(output, "human_F49B7.seurat.train.rds"))
saveRDS(human_IMR90.seurat.train, file = paste0(output, "human_IMR90.seurat.train.rds"))
# rerun check test and then save test data
saveRDS(human_F49B7.seurat.test, file = paste0(output, "human_F49B7.seurat.test.rds"))
saveRDS(human_IMR90.seurat.test, file = paste0(output, "human_IMR90.seurat.test.rds"))
```
# if you need to read objects back in
```{r read_in}
human_D205.seurat.train <- readRDS(file = "human_D205_models.rds")
gamm <- readRDS(file = "output_20230711_155556/GAMM_S2_norm.rds")
```
# classify cells
```{r classify_cells}
# An important requirement for classifying cells is using the same normalization 
# method for both the reference and the query datasets.
# query data was normalized in preprocess_crossspecies.Rmd
# scPred now uses Harmony to align the query data onto the training low-dimensional 
# space used as reference. Once the data is aligned, cells are classified using the 
# pre-trained models.
# predict query
gamms2.1 <- scPredict(gamms2, human_F49B7.seurat.train)
gamms2.2 <- scPredict(gamms2, human_IMR90.seurat.train)
#plot the classifications over the aligned data.
pdf(paste0(output, "celltype_prediction_pca_gamms2.1.pdf"), width = 8, height = 6)
DimPlot(gamms2.1, group.by = "scpred_prediction", reduction = "scpred")
dev.off()
pdf(paste0(output, "celltype_prediction_pca_gamms2.2.pdf"), width = 8, height = 6)
DimPlot(gamms2.2, group.by = "scpred_prediction", reduction = "scpred")
dev.off()
# run UMAP using the aligned data as an input
gamms2.1 <- RunUMAP(gamms2.1, reduction = "scpred", dims = 1:30)
gamms2.2 <- RunUMAP(gamms2.2, reduction = "scpred", dims = 1:30)
# plot the predicted labels for each cell type over the UMAP:
pdf(paste0(output, "celltype_prediction_umap_gamms2.1.pdf"), width = 8, height = 6)
DimPlot(gamms2.1, group.by = "scpred_prediction", label = TRUE, repel = TRUE)
dev.off()
pdf(paste0(output, "celltype_prediction_umap_gamms2.2.pdf"), width = 8, height = 6)
DimPlot(gamms2.2, group.by = "scpred_prediction", label = TRUE, repel = TRUE)
dev.off()
```
# visualize probabilities
```{r visualize_prob}
#  probabilities of each cell in the @meta.data slot of the query Seurat object.
# get labels
labels<- as.vector(unique(colnames(gamms2.1@meta.data)))
labels<-labels[! labels %in% c('orig.ident', 'nCount_RNA', 'nFeature_RNA', 
        'scpred_max', 'scpred_prediction', 'scpred_no_rejection','sizeFactors',
        "scDblFinder.score","percent.mt","CellType","ident","scDblFinder.weighted",
        "RNA_snn_res.0.5","scDblFinder.class","scDblFinder.cxds_score","seurat_clusters",
        'S.Score','G2M.Score','Phase','old.ident')]
# visualize the probabilities over the UMAP plot:
pdf(paste0(output, "celltype_prediction_featureplot_gamms2.1.pdf"), width = 11, height = 9)
print(FeaturePlot(gamms2.1, labels, keep.scale = "all"))
dev.off()
# get labels
labels<- as.vector(unique(colnames(gamms2.2@meta.data)))
labels<-labels[! labels %in% c('orig.ident', 'nCount_RNA', 'nFeature_RNA', 
        'scpred_max', 'scpred_prediction', 'scpred_no_rejection','sizeFactors',
        "scDblFinder.score","percent.mt","CellType","ident","scDblFinder.weighted",
        "RNA_snn_res.0.5","scDblFinder.class","scDblFinder.cxds_score","seurat_clusters",
        'S.Score','G2M.Score','Phase','old.ident')]
pdf(paste0(output, "celltype_prediction_featureplot_gamms2.2.pdf"), width = 11, height = 9)
print(FeaturePlot(gamms2.2, labels, keep.scale = "all"))
dev.off()
```
# verify performance with manual annotation
```{r read in metadata}
# add in metadata from previous analysis
metadata.gamm <- read.csv(METADATA, row.names = 1, 
            header = TRUE, sep = "\t")
# check metadata with query data
colnames(gamm@assays$RNA@data)[1:10]
rownames(metadata.gamm)[1:10]
# remove "_1" from metadata
rownames(metadata.gamm) <- sub("_1", "", rownames(metadata.gamm))
rownames(metadata.gamm)[1:10]
# add metadata to query data
gamms2.1 <- AddMetaData(gamms2.1, metadata.gamm)
gamms2.2 <- AddMetaData(gamms2.2, metadata.gamm)
```

# verify model performance in query data
```{r verify_performance}
query.ref.data.table<- as.data.frame(crossTab(gamms2.1, "CellType", "scpred_prediction"))
# check out proportion of cells
query.ref.data.table2<- as.data.frame(crossTab(gamms2.1, "CellType", "scpred_prediction", output = "prop"))
query.ref.data.table3<- as.data.frame(crossTab(gamms2.2, "CellType", "scpred_prediction"))
# check out proportion of cells
query.ref.data.table4<- as.data.frame(crossTab(gamms2.2, "CellType", "scpred_prediction", output = "prop"))
write.table(query.ref.data.table, file = paste0(output, "query2.1_ref_data_table_0.5.txt"), sep = "\t")
write.table(query.ref.data.table2, file = paste0(output, "query2.1_ref_data_table_proportion_0.5.txt"), sep = "\t")
write.table(query.ref.data.table3, file = paste0(output, "query2.2_ref_data_table_0.5.txt"), sep = "\t")
write.table(query.ref.data.table4, file = paste0(output, "query2.12ref_data_table_proportion_0.5.txt"), sep = "\t")
# get query umap with orignal labels
pdf(paste0(output, "query_celltype_originallabels2.1-umap_0.5.pdf"), width = 8, height = 6)
DimPlot(gamms2.1, group.by = "CellType", repel = TRUE, label=TRUE)#+ NoLegend()
dev.off()
pdf(paste0(output, "query_celltype_originallabels2.2-umap_0.5.pdf"), width = 8, height = 6)
DimPlot(gamms2.2, group.by = "CellType", repel = TRUE, label=TRUE)#+ NoLegend()
dev.off()
```
# save query seurat object
```{r save_query}
saveRDS(gamms2.1, file = paste0(output, "GAMM_S2.1_scpred_c0.5.rds"))
saveRDS(gamms2.2, file = paste0(output, "GAMM_S2.2_scpred_c0.5.rds"))
```
```{r access_classifiers}
# access the classifiers
classifiers<- get_classifiers(human_F49B7.seurat.train)
classifiers2<- get_classifiers(human_IMR90.seurat.train)
saveRDS(classifiers, file = paste0(output, "classifiersF49B7.rds"))
saveRDS(classifiers2, file = paste0(output, "classifiersIMR90.rds"))
```
# save session info
```{r sessioninfo}
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```