---
title: "sc Pred"
author: "Beth Moore"
date: "2023-09-18"
output: html_document
---
# load packages
```{r load_packages}
knitr::opts_chunk$set(echo = TRUE)
# conda activate scRNAseq_best
#devtools::install_github("powellgenomicslab/scPred")
library(scPred)
library(Seurat)
library(magrittr)
library(harmony)
library(rmarkdown)
library(jsonlite)
library(purrr)
library(scran)
library(patchwork)
library(dplyr)
library(reticulate)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/"
METADATA <- "gamm_manual_annot_metadata_c0.5.txt" # add to working directory
use_condaenv(condaenv = 'scRNAseq_best', required = TRUE)
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/reh_cellrep_2020/"))
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_scPred_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json"))
config <- fromJSON(file.path(output, "config.json"))
output <- paste0(output, "/")
source("/w5home/bmoore/scRNAseq_library/sc_pipeline/src/sc_pipeline_functions.R")
```
# run preprocess_crossspecies.Rmd first to get the data preprocessed the same way
# get reference and query data
```{r get_data}
#human_D205.seurat <- readRDS(file = "output_20230711_155556/human_D205_norm.rds")
human_D205.seurat <- readRDS(file = "output_seurat_mapping_20231110_135152/human_D205_clustered.rds")
#gamm <- readRDS(file = "output_20230711_155556/GAMM_S2_norm.rds")
gamm <- readRDS(file = "output_seurat_mapping_20231110_135152/gamm.S2_clustered.rds")
```
# subset metadata if needed
```{r subset_metadata}
# check metadata
table(human_D205.seurat$type)
# subset if needed
human_D205.seurat<- subset(human_D205.seurat, subset = type != c("Midbrain"))
human_D205.seurat<- subset(human_D205.seurat, subset = type != c('T2'))
human_D205.seurat<- subset(human_D205.seurat, subset = type != c('AC2'))
human_D205.seurat<- subset(human_D205.seurat, subset = type != c('miG'))
table(human_D205.seurat$type)
```
```{r subset_test_set}
# Set the percentage of cells to sample (10% in this case)
sampling_percentage <- 0.1
# Calculate the number of cells to sample for each cell type
cell_types <- unique(human_D205.seurat$type)
cells_to_sample <- ceiling(table(human_D205.seurat$type) * sampling_percentage)

# Create a copy of the original Seurat object to keep the unsubsetted cells
human_D205.seurat.train <- human_D205.seurat

# Initialize an empty list to store subsetted Seurat objects
subset_list <- list()

# Iterate through each cell type, subset the cells, and store the subset in the list
for (cell_type in cell_types) {
  cell_indices <- which(human_D205.seurat$type == cell_type)
  sampled_indices <- sample(cell_indices, cells_to_sample[cell_type], replace = FALSE)
  
  subset_list[[cell_type]] <- human_D205.seurat[, sampled_indices]
  # Remove the subsetted cells from the original Seurat object
  human_D205.seurat.train <- human_D205.seurat.train[, -sampled_indices]
}

# Create a new Seurat object with the subsetted cells
human_D205.seurat.test <- CreateSeuratObject(counts = subset_list[[1]]@assays$RNA@counts,
                                           meta.data = subset_list[[1]]@meta.data)
for (i in 2:length(subset_list)) {
  human_D205.seurat.test <- merge(human_D205.seurat.test, subset_list[[i]])
}

human_D205.seurat.test$Idents <- human_D205.seurat.test$type
```
# Perform further analysis as needed on the seurat_test_object
```{r normalize_test_set}
human_D205.seurat.test_norm <- normalize_data(human_D205.seurat.test)
#saveRDS(human_D205.seurat.test_norm, file= paste0(output,"human_D205_test_norm.rds"))
```

# do analysis on training set
```{r redo_analysis}
# no need to normalize again
#feature selection
human_D205.seurat.train <- feature_selection(human_D205.seurat.train)
# scale
human_D205.seurat.train <- scale_data(human_D205.seurat.train)
# run pca
human_D205.seurat.train <- run_and_visualize_pca(human_D205.seurat.train)
# run umap
human_D205.seurat.train <- run_umap(human_D205.seurat.train)
#save
#saveRDS(human_D205.seurat.train, file= paste0(output,"human_D205_train.rds"))
```
# show umap of ref data
```{r show_umap_ref}
# we don't do clustering because clusters already annotated
 # Save UMAP clusters plot
  pdf(paste0(output, "umap_ref.pdf"), width = 8, height = 6)
  umap_clusters <- DimPlot(human_D205.seurat.train, group.by = "type", 
  reduction = "umap", label = TRUE, pt.size = .1)
  print(umap_clusters)
  dev.off()
  ```
# training the classifyer
```{r train_classifyer}
# getFeatureSpace will create a scPred object stored in the @misc slot. This 
# object will contained all required information to classify cells.
human_D205.seurat.train <- getFeatureSpace(human_D205.seurat.train, "type")

# train models for each cell type
# cv has to be set BELOW min class
human_D205.seurat.train <- trainModel(human_D205.seurat.train, resampleMethod = "cv",
                                number = 10, seed = 704)
# if you get this error:
# There were missing values in resampled performance measures.
# try: 1. make sure no NAs in data 2. decreasing the cvs, 3. down or up sample to make classes even
```
```{r check_classifyer}
# get training prob for each cell-cell type
get_probabilities(human_D205.seurat.train) %>% head()
# use get_scpred method to retrieve the scPred object from the Seurat object.
# this gives stats on each cell type prediction
get_scpred(human_D205.seurat.train)
model_stats <- get_scpred(human_D205.seurat.train)
model_stats
# visualize cell type probabilties
pdf(paste0(output, "celltype_model_prob_svmradial.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat.train)
dev.off()
```
# OPTIONAL: can try other models from caret:
# https://topepo.github.io/caret/available-models.html
# pass in the model parameter
# can reclassify a subset of cells (like those that didn't work well)

# check test data
```{r check_test}
human_D205.seurat.test <- scPredict(human_D205.seurat.test, human_D205.seurat.train)
# test set in predictions space
pdf(paste0(output, "test_celltype_prediction_fin.pdf"), width = 8, height = 6)
DimPlot(human_D205.seurat.test, group.by = "scpred_prediction", reduction = "scpred")
dev.off()
# run UMAP using the aligned data as an input
human_D205.seurat.test <- RunUMAP(human_D205.seurat.test, reduction = "scpred", dims = 1:30)
# plot the predicted labels for each cell type over the UMAP:
pdf(paste0(output, "test_celltype_prediction_fin-umap.pdf"), width = 8, height = 6)
DimPlot(human_D205.seurat.test, group.by = "scpred_prediction", label = TRUE, repel = TRUE)
dev.off()
# compare to original labels
pdf(paste0(output, "test_celltype_original-label-umap-fin.pdf"), width = 8, height = 6)
DimPlot(human_D205.seurat.test, group.by = "type", label = TRUE, repel = TRUE)
dev.off()
```
# can reclassify a subset of cells (like those that didn't work well)
```{r train_classifyer2}
# other models tried: rf, logreg, xgbTree, mda, AdaBoost.M1
human_D205.seurat.train <- trainModel(human_D205.seurat.train, model = 'AdaBoost.M1', 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("Midbrain","T1/T3","Prog/Glia","RGC"))
get_scpred(human_D205.seurat.train)
model_stats_ada <- get_scpred(human_D205.seurat.train)
pdf(paste0(output, "celltype_model_prob_adaboost.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat.train)
dev.off()
# note run check test again to see if model improved on test data
```
```{r train_classifyer3}
# other models 'glm'
human_D205.seurat.train <- trainModel(human_D205.seurat.train, model = "glm", 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("Midbrain","T1/T3","RGC"))
get_scpred(human_D205.seurat.train)
pdf(paste0(output, "celltype_model_prob_glm.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat.train)
dev.off()
# run test data again
```
# get final model
```{r train_classifyer4}
# reclassify prog/glia and rgc with svm-radial
human_D205.seurat.train <- trainModel(human_D205.seurat.train, model = "svmRadial", 
                                   resampleMethod = "cv",
                                   number = 10, seed = 704,
                                   reclassify = c("Midbrain","RGC"))
get_scpred(human_D205.seurat.train)
pdf(paste0(output, "celltype_model_prob_fin.pdf"), width = 8, height = 6)
plot_probabilities(human_D205.seurat.train)
dev.off()
```
# save seurat object
```{r save_ref}
# save ref data
saveRDS(human_D205.seurat.train, file = paste0(output, "human_D205_train_models.rds"))
# rerun check test and then save test data
saveRDS(human_D205.seurat.test, file = paste0(output, "human_D205_test.rds"))
```
# if you need to read objects back in
```{r read_in}
human_D205.seurat.train <- readRDS(file = "output_scPred_20230913_131047_filt/human_D205_train_models.rds")
gamm <- readRDS(file = "output_preprocess20230912_144047_cc/GAMM_S2_norm.rds")
```
# classify cells
```{r classify_cells}
# An important requirement for classifying cells is using the same normalization 
# method for both the reference and the query datasets.
# query data was normalized in preprocess_crossspecies.Rmd
# scPred now uses Harmony to align the query data onto the training low-dimensional 
# space used as reference. Once the data is aligned, cells are classified using the 
# pre-trained models.
# predict query
gamm <- scPredict(gamm, human_D205.seurat.train)

#plot the classifications over the aligned data.
pdf(paste0(output, "celltype_prediction_1_gamms2.pdf"), width = 8, height = 6)
DimPlot(gamm, group.by = "scpred_prediction", reduction = "scpred")
dev.off()
# run UMAP using the aligned data as an input
gamm <- RunUMAP(gamm, reduction = "scpred", dims = 1:30)
# plot the predicted labels for each cell type over the UMAP:
pdf(paste0(output, "celltype_prediction_1-umap_gamms2.pdf"), width = 8, height = 6)
DimPlot(gamm, group.by = "scpred_prediction", label = TRUE, repel = TRUE)
dev.off()
```
# visualize probabilities
```{r visualize_prob}
#  probabilities of each cell in the @meta.data slot of the query Seurat object.
# get labels
labels<- as.vector(unique(colnames(gamm@meta.data)))
labels<-labels[! labels %in% c('orig.ident', 'nCount_RNA', 'nFeature_RNA', 
        'scpred_max', 'scpred_prediction', 'scpred_no_rejection','sizeFactors',
        "scDblFinder.score","percent.mt","CellType","ident","scDblFinder.weighted",
        "RNA_snn_res.0.5","scDblFinder.class","scDblFinder.cxds_score","seurat_clusters",
        'S.Score','G2M.Score','Phase','old.ident')]
# visualize the probabilities over the UMAP plot:
pdf(paste0(output, "celltype_prediction_featureplot_gamms2.pdf"), width = 11, height = 9)
print(FeaturePlot(gamm, labels, keep.scale = "all"))
dev.off()
```
# verify performance with manual annotation
```{r read in metadata}
# add in metadata from previous analysis
metadata.gamm <- read.csv(METADATA, row.names = 1, 
            header = TRUE, sep = "\t")
# check metadata with query data
colnames(gamm@assays$RNA@data)[1:10]
rownames(metadata.gamm)[1:10]
# remove "_1" from metadata
rownames(metadata.gamm) <- sub("_1", "", rownames(metadata.gamm))
rownames(metadata.gamm)[1:10]
# add metadata to query data
gamm <- AddMetaData(gamm, metadata.gamm)
```

# verify model performance in query data
```{r verify_performance}
query.ref.data.table<- as.data.frame(crossTab(gamm, "CellType", "scpred_prediction"))
# check out proportion of cells
query.ref.data.table2<- as.data.frame(crossTab(gamm, "CellType", "scpred_prediction", output = "prop"))
# order by vector
rowvec <- c("BC","PR","iMG","Prog/Glia","Prog","RGC","AC","HC","T1/T3","unassigned")
colvec <- c("Bipolar Cells - Ganglion Cells","Pan PRs","Pan PRs - Rods","Cones - Pan PRs","Muller Glia",
"Retinal Prog - Muller Glia","Rods - Pan PRs - Muller Glia","Ganglion Cells","unknown")
query.ref.data.table<- query.ref.data.table[order(match(rownames(query.ref.data.table), rowvec)), , drop = FALSE]
query.ref.data.table<- query.ref.data.table[colvec]
query.ref.data.table2<- query.ref.data.table2[order(match(rownames(query.ref.data.table2), rowvec)), , drop = FALSE]
query.ref.data.table2<- query.ref.data.table2[colvec]
# write out
write.table(query.ref.data.table, file = paste0(output, "query_ref_data_table_0.5.txt"), sep = "\t")
write.table(query.ref.data.table2, file = paste0(output, "query_ref_data_table_proportion_0.5.txt"), sep = "\t")
# get query umap with orignal labels
pdf(paste0(output, "query_celltype_originallabels-umap_0.5.pdf"), width = 8, height = 6)
DimPlot(gamm, group.by = "CellType", repel = TRUE, label=TRUE)#+ NoLegend()
dev.off()
```
# make heatmap
```{r heatmap}
heatmap_func <- function(df){
  # turn df into matrix
  mat <- as.matrix(sapply(df, as.numeric, rownames=TRUE))  
  y <- rownames(df)
  rownames(mat) <- y
  print(df[1,1])
  print(mat[1,1])

  # make heatmap
  hm<- Heatmap(mat, name= "Proportion of cells", column_title = "Cell types - manual", 
             row_title = "Cell types - predicted", column_title_side = "top",row_title_side = "left", 
             col = colorRamp2(c(0, 0.5, 1), c("blue","white","darkred")),
             cluster_columns = F, cluster_rows= F, show_row_dend = F, 
             column_names_gp = gpar(fontsize = 10), show_column_names = T, 
             show_row_names = T, row_names_gp = gpar(fontsize = 10), 
             cell_fun = function(j, i, x, y, width, height, fill) {
               grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 8))})
  return(hm)
}
hm <- heatmap_func(query.ref.data.table2)
pdf(file = paste0(output,"celltype_manualvs.predicted_heatmap.pdf"), width = 7, height = 5)
print(hm)
dev.off()
```
# save query seurat object
```{r save_query}
saveRDS(gamm, file = paste0(output, "GAMM_S2_scpred_c0.5.rds"))
```
```{r access_classifiers}
# access the classifiers
classifiers<- get_classifiers(human_D205.seurat.train)

saveRDS(classifiers, file = paste0(output, "classifiers.rds"))
```
# save session info
```{r sessioninfo}
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```