---
title: "seurat_integration_within_species"
author: "Beth Moore"
date: "2023-12-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(devtools)
library(Seurat)
# devtools::install_github('satijalab/seurat-data')
library(SeuratData)
library(patchwork)
library(purrr)
library(jsonlite)
library(rmarkdown)
library(reticulate)
library(multtest)
library(metap)
BETHS_GAMM_DATA_DIR <- "/w5home/bmoore/scRNAseq/GAMM/"
use_condaenv("scRNAseq_best")
setwd(paste0(BETHS_GAMM_DATA_DIR, "human_data/cowan_cell_2020/"))
timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
output <- paste0("output_seurat_integrate_", timestamp)
dir.create(output, showWarnings = FALSE)
file.copy("/w5home/bmoore/scRNAseq_library/config.json", file.path(output, "config.json"))
#output <- "output_seurat_20230731_094648/"
cell.cycle.markers.s <- read.csv2("/w5home/bmoore/scRNAseq_library/sc_pipeline/cell_cycle_vignette/cell_cycle_orthologs_s.genes.txt", 
    sep = "\t", header = TRUE, row.names = 1)
cell.cycle.markers.g2m <- read.csv2("/w5home/bmoore/scRNAseq_library/sc_pipeline/cell_cycle_vignette/cell_cycle_orthologs_g2m.genes.txt", 
    sep = "\t", header = TRUE, row.names = 1)
varslist <- c(cell.cycle.markers.s, cell.cycle.markers.g2m)
output <- paste0(output, "/")
config <- fromJSON(file.path(output, "config.json"))
source("/w5home/bmoore/scRNAseq_library/sc_pipeline/src/sc_pipeline_functions.R")
```
# if needed, install python packages
```{r}
library(reticulate)
py_install("umap-learn")
py_install("leidenalg")
py_install("pandas")
```
# other packages
```{r}
install.packages('BiocManager')
BiocManager::install('multtest')
install.packages('metap')
```

# load data
```{r load_data}
gamms2<- readRDS(file = "~/scRNAseq/GAMM/human_data/reh_cellrep_2020/output_preprocess20230808_113032/GAMM_S2_feat_select.rds")
human_D205.seurat<- readRDS(file = "~/scRNAseq/GAMM/human_data/reh_cellrep_2020/output_preprocess20230808_113032/human_D205_feat_select.rds")
```
# select features for integration across specues
```{r select_features}
# create list of seurat objects
object.list <- list(gamms2, human_D205.seurat)
# select features
features <- SelectIntegrationFeatures(object.list = object.list)
```
# integrate data
```{r integrate_data}
# find anchors
gamm.anchors <- FindIntegrationAnchors(object.list = object.list, anchor.features = features)
# integrate data
# this command creates an 'integrated' data assay
gamm.combined <- IntegrateData(anchorset = gamm.anchors)
```
```{r set_default_assay}
DefaultAssay(gamm.combined) <- "integrated"
```

# scale, dimensionality reduction, and clustering on combined data
```{r scale}
# scale data
gamm.combined<- scale_data(gamm.combined, vars_list = varslist, path = output)
```
```{r run_dim_reduction}
# run PCA
gamm.combined<- run_and_visualize_pca(gamm.combined)
# run UMAP
gamm.combined<- run_umap(gamm.combined)
```
```{r run_clustering}
# run clustering
## Note: you must have the reduction in the config file set to pca in order for this to work
gamm.combined<- perform_clustering(gamm.combined)
```
```{r save_data}
saveRDS(gamm.combined, file= paste0(output,"gamm.combined_cc.regress_1to1.rds"))
```
```{r load_data}
gamm.combined<- readRDS(file = "/w5home/bmoore/scRNAseq/GAMM/human_data/reh_cellrep_2020/output_seurat_20230731_094648/gamm.combined.rds")
```

# Visualization
```{r}
p1 <- DimPlot(gamm.combined, reduction = "umap", group.by = "orig.ident")
p2 <- DimPlot(gamm.combined, reduction = "umap", label = TRUE, repel = TRUE)
pdf(paste0(output, "seurat_integration-umap.pdf"), width = 11, height = 6)
print(p1 + p2)
dev.off()
p3 <- DimPlot(gamm.combined, reduction = "pca", group.by = "orig.ident")
p4 <- DimPlot(gamm.combined, reduction = "pca", label = TRUE, repel = TRUE)
pdf(paste0(output, "seurat_integration-pca.pdf"), width = 11, height = 6)
print(p3 + p4)
dev.off()
pdf(paste0(output, "seurat_integration-umap2.pdf"), width = 8, height = 6)
DimPlot(gamm.combined, reduction = "umap", split.by = "orig.ident")
dev.off()
pdf(paste0(output, "seurat_integration-pca2.pdf"), width = 8, height = 6)
DimPlot(gamm.combined, reduction = "pca", split.by = "orig.ident")
dev.off()
```
# Identify conserved markers
```{r}
# For performing differential expression after integration, we switch back to the original
# data
DefaultAssay(gamm.combined) <- "RNA"
#markers.7 <- FindConservedMarkers(gamm.combined, ident.1 = 7, grouping.var = "orig.ident", verbose = FALSE)
#head(markers.7)
```
# order and subset markers
```{r}
ordered.markers.1 <- markers.7[order(markers.7$max_pval, decreasing = FALSE),]
new_df.ordered <- subset(ordered.markers.1, D205_avg_log2FC >= 0.5 & 
    gamm_S2_avg_log2FC >= 0.5 & D205_p_val_adj < 0.05 & gamm_S2_p_val_adj < 0.05)
as.data.frame(new_df.ordered) %>% write.csv(file = paste0(output, "DEgenes.7.csv"))
```
# read in known markers
```{r}
known.markers <- read.csv2("/w5home/bmoore/scRNAseq_library/known_marker_lists/Gamm_lab_Consolidated_markerList.txt", sep = "\t", header = TRUE, row.names = 1)
known.markers.df <- as.data.frame(known.markers)
```
# merge known markers with new markers
```{r}
marker_df <- merge(as.data.frame(new_df.ordered), known.markers.df, by = 'row.names')
marker_df.ordered <- marker_df[order(marker_df$max_pval, decreasing = FALSE),]
write.csv(marker_df.ordered, file = paste0(output, "markers.1.csv"))
top10 <- head(marker_df.ordered, 10)
```
# plot marker genes
```{r}
pdf(paste0(output, "marker_gene_feature_plot_clust1.pdf"), width = 10, height = 8)
print(FeaturePlot(gamm.combined, features = top10$Row.names, 
    min.cutoff = "q9"))
dev.off()
```
# for loop to get conserved markers for each cluster
```{r}
for (i in levels(gamm.combined)){
    print(i)
    # find DE genes
    de.genes <- FindConservedMarkers(gamm.combined, ident.1 = strtoi(i), grouping.var = "orig.ident", verbose = FALSE)
    # order and subset
    ordered.de.genes <- de.genes[order(de.genes$max_pval, decreasing = FALSE),]
    new_df.ordered <- subset(ordered.de.genes, D205_avg_log2FC >= 0.5 & 
    gamm_S2_avg_log2FC >= 0.5 & D205_p_val_adj < 0.05 & gamm_S2_p_val_adj < 0.05)
    as.data.frame(new_df.ordered) %>% write.csv(file = paste0(output, "DEgenes.", i, ".csv"))
    # merge with known markers
    marker_df <- merge(as.data.frame(new_df.ordered), known.markers.df, by = 'row.names')
    marker_df.ordered <- marker_df[order(marker_df$max_pval, decreasing = FALSE),]
    write.csv(marker_df.ordered, file = paste0(output, "markers.", i, ".csv"))
    
    # check for empty dataframe
    if (nrow(marker_df) == 0){
      print(paste0("This cluster does not have any markers: ", i))
    }
    else {
    # get top 10
    top10 <- head(marker_df.ordered, 10)
    # plot marker genes
    pdf(paste0(output, "marker_gene_feature_plot_clust",i,".pdf"), width = 10, height = 8)
    print(FeaturePlot(gamm.combined, features = top10$Row.names, 
    min.cutoff = "q9"))
    dev.off()
}}
```
# annotate clusters
```{r}
gamm.combined <- RenameIdents(gamm.combined, `1` = "Amacrine-Muller Glia", 
    `2` = "none", `3` = "none",`4` = "none", `5` = "none", 
    `6` = "none", `7` = "Pan PRs", `8` = "Cone", 
    `9` = "none", `10` = "none", `11` = "none", `12` = "Cone"
    )
pdf(paste0(output, "combined_umap_label.pdf"), width = 8, height = 6)
print(DimPlot(gamm.combined, label = TRUE))
dev.off()
```
```{r save_data}
saveRDS(gamm.combined, file= paste0(output,"gamm.combined.rds"))
```
# make dot plot
```{r}
markers.to.plot <- c("SFRP2","CCND1","COL2A1","VIM","RLBP1","RCVRN","AIPL1","SLC4A10",
    "NRXN3","GFAP","AQP4","CLU","SLC1A3","EZR","GNGT1","ROBO2","NETO1","CA10")
pdf(paste0(output, "combined_dotplot.pdf"), width = 8, height = 8)
DotPlot(gamm.combined, features = markers.to.plot, cols = c("blue", "red"), dot.scale = 8, 
    split.by = "orig.ident") +RotatedAxis()
dev.off()
```
# identify DE genes across species
```{r}
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
# get cell types that match up across species
rprog.mg.cells <- subset(gamm.combined, idents = "Retinal Prog-Muller Glia")
Idents(rprog.mg.cells) <- "orig.ident"
Idents(rprog.mg.cells)
# get average expression for Retinal Prog-Muller Glia genes
avg.rprog.mg.cells <- as.data.frame(log1p(AverageExpression(rprog.mg.cells, verbose = FALSE)$RNA))
avg.rprog.mg.cells$gene <- rownames(avg.rprog.mg.cells)
# get marker genes to label
genes.to.label = c("SFRP2","CCND1","COL2A1","VIM","RLBP1","SLC1A3",
"DKK3","FZD5","SPON1","FABP7","SIX6","HES1","PMEPA1","EPHB2","SIX3",
"LGALS1","IGFBP5")
# plot
p1 <- ggplot(avg.rprog.mg.cells, aes(D205, gamm_S2)) + geom_point() + ggtitle("Retinal Prog-Muller Glia")
p1 <- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)
pdf(paste0(output,"DEgenes_RprogMG.pdf"), width = 8, height = 6)
print(p1)
dev.off()
```
# find markers between 2 species
```{r}
gamm.combined$celltype.orig.ident <- paste(Idents(gamm.combined), gamm.combined$orig.ident, sep = "_")
gamm.combined$celltype <- Idents(gamm.combined)
Idents(gamm.combined) <- "celltype.orig.ident"
levels(gamm.combined)
rprog.mg.diff <- FindMarkers(gamm.combined, ident.1 = "Retinal Prog-Muller Glia_gamm_S2", ident.2 = "Retinal Prog-Muller Glia_D205", verbose = FALSE)
head(rprog.mg.diff, n = 15)
```
```{r}
rprog.mg.diff.ordered <- rprog.mg.diff[order(rprog.mg.diff$p_val_adj, -abs(rprog.mg.diff$avg_log2FC)),]
write.csv(rprog.mg.diff.ordered, file = paste0(output, "rprog.mg.diff.csv"))
```
# plot markers top 5
```{r}
top5 <- head(rprog.mg.diff.ordered, 5)
pdf(paste0(output, "rprog.mg.diff_top5_featureplot.pdf"), width = 6, height = 10)
print(FeaturePlot(gamm.combined, features = rownames(top5), min.cutoff = "q9",
    cols = c("grey", "red"), split.by = "orig.ident")) #, max.cutoff=3
dev.off()
plots <- VlnPlot(gamm.combined, features = rownames(top5), split.by = "orig.ident", 
    group.by = "celltype", pt.size = 0, combine = FALSE)
pdf(paste0(output, "rprog.mg.diff_top5_vlnplot.pdf"), width = 8, height = 14)
print(wrap_plots(plots = plots, ncol = 1))
dev.off()
```
# save session info
```{r sessioninfo}
writeLines(capture.output(sessionInfo()), paste0(output,"sessionInfo.txt"))
```